File: ts/src/kis/kis-candle-service.ts
/**
 * @fileoverview KIS 가 직접 주는 캔들(OHLCV) 조회
 * @description `kis.fetchOHLCV` 는 이력이 충분한 야후 파이낸스를 먼저 쓴다. 이 모듈은 KIS 원본 캔들이 필요한 경로(심층 이력 채우기, 당일 분봉,
 * 야후가 빈 해외 일봉)를 `kis` 인스턴스의 암묵 API 로 조회한다.
 *
 * - fetchDailyOHLCV: 기간별(일/주/월) 캔들
 * - fetchDailyOHLCVPaged: 창을 과거로 옮겨 가며 심층 이력을 모은다
 * - fetchMinuteOHLCV: 당일 분봉
 * - fetchOverseasDailyOHLCV: 미국 기간별(일/주/월) 캔들
 * - resampleMinuteCandles: N분봉 리샘플링
 */

import { logger } from '../logger';
import { etYmd } from '../us-market-hours';
import { resampleCandles, type OhlcvRow } from './candle-resample';
import type { Exchange } from '../base';
import { planWindows, mergeCandles, sliceCandleWindow, toKisDate } from './kis-candle-pagination';
import type { KisDailyCandle, KisOverseasDailyCandle } from './kis-types';
import type { OverseasMarket } from './kis-overseas-master';

// ============ 캔들 관련 상수 ============

/** KIS 캔들 조회용 거래 ID */
const CANDLE_TR_IDS = {
    /** 기간별 시세 (일/주/월/년) — 국내 */
    DAILY_CHART: 'FHKST03010100',
    /** 분봉 시세 (당일) — 국내 */
    MINUTE_CHART: 'FHKST03010200',
    /** 해외주식 기간별시세 (일/주/월) — overseas-price/v1/quotations/dailyprice */
    OVERSEAS_DAILY_CHART: 'HHDFS76240000',
} as const;

/** 해외주식 일/주/월 구분 코드 (GUBN) */
const OVERSEAS_GUBN_MAP: Record<string, '0' | '1' | '2'> = {
    '1d': '0',
    '1w': '1',
    '1W': '1',
    '1M': '2',
};

/** 분봉 API 1회 최대 반환 건수 */
const MINUTE_PAGE_SIZE = 30;

/** 기간 코드별 일수 승수 (주말/공휴일 마진 포함) */
const PERIOD_DAY_MULTIPLIER: Record<string, number> = {
    'W': 7,
    'M': 30,
};

/** 날짜 마진 계수 (공휴일/주말 고려) */
const DATE_MARGIN_FACTOR = 1.5;

/** 해외 일/주/월봉 한 개가 차지하는 달력일 수. `since` 에서 `limit` 개를 덮는 기준일을 어림할 때 쓴다. */
const OVERSEAS_PERIOD_DAYS: Record<string, number> = {
    '1d': 1,
    '1w': 7,
    '1W': 7,
    '1M': 31,
};

/** 기준일 어림에 더하는 달력일. `limit` 이 작아도 연휴를 덮는다. */
const OVERSEAS_END_PAD_DAYS = 7;

const DAY_MS = 24 * 60 * 60 * 1000;

// ============ 서비스 ============

export class KisCandleService {
    /** @param exchange 자격증명을 채운 `kis` 인스턴스. 암묵 API(`privateGet…`)를 부른다. */
    constructor(private readonly exchange: Exchange) { }

    /**
     * 기간별(일/주/월) 캔들 조회
     * KIS inquire-daily-itemchartprice 사용. 조회 기간은 실행 환경의 시간대가 아니라 한국 날짜로 적는다.
     */
    async fetchDailyOHLCV(
        stockCode: string,
        periodCode: string,
        limit: number,
    ): Promise<number[][]> {
        const now = this.exchange.milliseconds();
        const dayMultiplier = PERIOD_DAY_MULTIPLIER[periodCode] ?? 1;
        const startDate = toKisDate(new Date(now - limit * dayMultiplier * DATE_MARGIN_FACTOR * DAY_MS));
        const rows = await this.fetchDailyOHLCVRange(stockCode, periodCode, startDate, toKisDate(new Date(now)));
        return rows.slice(-limit);
    }

    /**
     * 창을 과거로 옮겨 가며 **여러 번 호출해** 심층 이력을 채운다.
     *
     * KIS 한 응답이 ~100행이라, 800봉을 원하면 8번 호출한다.
     *
     * **빈 창이 나오면 멈춘다.** 상장 이전 구간까지 창을 계속 옮기면 호출만 낭비된다.
     * 신규 상장주는 1~2창에서 끝난다.
     *
     * @param neededCandles 더 받아야 할 캔들 수. 0 이하면 **한 번도 호출하지 않는다**.
     */
    async fetchDailyOHLCVPaged(
        stockCode: string,
        periodCode: string,
        neededCandles: number,
        opts: { nowMs?: number; onPage?: (i: number) => Promise<void> } = {},
    ): Promise<number[][]> {
        const windows = planWindows(neededCandles, opts.nowMs ?? this.exchange.milliseconds());
        if (windows.length === 0) return [];

        const pages: number[][][] = [];
        for (const [i, w] of windows.entries()) {
            const rows = await this.fetchDailyOHLCVRange(stockCode, periodCode, w.start, w.end);
            if (rows.length === 0) {
                logger.debug({ stockCode, page: i + 1, window: w },
                    '[KISCandleService] 빈 창 — 페이지네이션 중단(상장 이전 추정)');
                break;
            }
            pages.push(rows);
            if (opts.onPage) await opts.onPage(i);
        }
        const merged = mergeCandles(pages);
        logger.info({ stockCode, periodCode, pages: pages.length, candles: merged.length },
            '[KISCandleService] 페이지네이션 완료');
        return merged;
    }

    /**
     * 기간을 **명시해서** 일/주/월 캔들을 조회한다. 한 응답은 **약 100행**(`output2`)이라, 창을 옮겨 가며 부르려면 기간이 인자여야 한다.
     *
     * @param startDate `YYYYMMDD`
     * @param endDate `YYYYMMDD`
     */
    async fetchDailyOHLCVRange(
        stockCode: string,
        periodCode: string,
        startDate: string,
        endDate: string,
    ): Promise<number[][]> {
        try {
            const response = await this.exchange.privateGetUapiDomesticStockV1QuotationsInquireDailyItemchartprice({
                FID_COND_MRKT_DIV_CODE: 'J',
                FID_INPUT_ISCD: stockCode,
                FID_INPUT_DATE_1: startDate,
                FID_INPUT_DATE_2: endDate,
                FID_PERIOD_DIV_CODE: periodCode,
                FID_ORG_ADJ_PRC: '0',
                tr_id: CANDLE_TR_IDS.DAILY_CHART,
            });
            const data = response.output2 as KisDailyCandle[] | undefined;

            logger.info({
                stockCode, periodCode, startDate, endDate,
                length: Array.isArray(data) ? data.length : 'N/A',
            }, '[KISCandleService] fetchDailyOHLCV 응답');

            if (!Array.isArray(data)) return [];

            return data
                .map((candle): OhlcvRow => {
                    const dateStr = candle.stck_bsop_date;
                    const timestamp = new Date(
                        `${dateStr.slice(0, 4)}-${dateStr.slice(4, 6)}-${dateStr.slice(6, 8)}T09:00:00+09:00`
                    ).getTime();

                    return [
                        timestamp,
                        Number(candle.stck_oprc),
                        Number(candle.stck_hgpr),
                        Number(candle.stck_lwpr),
                        Number(candle.stck_clpr),
                        Number(candle.acml_vol),
                    ];
                })
                .sort((a, b) => a[0] - b[0]);
        } catch (err) {
            // 실패를 `[]` 로 바꾸지 않는다(분봉과 같다). 빈 창은 페이지네이션이 "상장 이전"으로 읽는다.
            logger.error({ err, stockCode, periodCode, startDate, endDate },
                '[KISCandleService] 기간별 캔들 조회 실패');
            throw err;
        }
    }

    /**
     * 분봉 캔들 조회 (당일만)
     * KIS inquire-time-itemchartprice 사용
     * 한 번에 최대 30건, 연속 조회로 limit까지 수집
     */
    async fetchMinuteOHLCV(
        stockCode: string,
        minuteInterval: number,
        limit: number,
    ): Promise<number[][]> {
        try {
            // 연속조회는 커서 시각의 봉을 다음 쪽에 다시 줄 수 있어, 시각마다 한 봉만 담는다.
            const byTimestamp = new Map<number, OhlcvRow>();
            let cursor = '';

            while (byTimestamp.size < limit) {
                const response = await this.exchange.privateGetUapiDomesticStockV1QuotationsInquireTimeItemchartprice({
                    FID_COND_MRKT_DIV_CODE: 'J',
                    FID_INPUT_ISCD: stockCode,
                    FID_INPUT_HOUR_1: cursor,
                    FID_PW_DATA_INCU_YN: 'N',
                    FID_ETC_CLS_CODE: '',
                    tr_id: CANDLE_TR_IDS.MINUTE_CHART,
                });
                const data = response.output2 as Array<Record<string, string>> | undefined;

                if (!Array.isArray(data) || data.length === 0) break;

                for (const item of data) {
                    const dateStr = item.stck_bsop_date;
                    const timeStr = item.stck_cntg_hour;
                    if (!dateStr || !timeStr) continue;

                    const timestamp = new Date(
                        `${dateStr.slice(0, 4)}-${dateStr.slice(4, 6)}-${dateStr.slice(6, 8)}` +
                        `T${timeStr.slice(0, 2)}:${timeStr.slice(2, 4)}:${timeStr.slice(4, 6)}+09:00`
                    ).getTime();

                    byTimestamp.set(timestamp, [
                        timestamp,
                        Number(item.stck_oprc),
                        Number(item.stck_hgpr),
                        Number(item.stck_lwpr),
                        Number(item.stck_prpr),
                        Number(item.cntg_vol),
                    ]);
                }

                if (data.length < MINUTE_PAGE_SIZE) break;

                const lastTime = data[data.length - 1]?.stck_cntg_hour;
                if (!lastTime || lastTime === cursor) break;
                cursor = lastTime;
            }

            logger.info({
                stockCode, minuteInterval, count: byTimestamp.size,
            }, '[KISCandleService] fetchMinuteOHLCV 완료');

            const sorted = [...byTimestamp.values()].sort((a, b) => a[0] - b[0]);
            if (minuteInterval <= 1) return sorted.slice(-limit);

            return this.resampleMinuteCandles(sorted, minuteInterval).slice(-limit);
        } catch (err) {
            // 실패를 `[]` 로 바꾸지 않는다. 빈 배열은 휴장·거래정지로 읽히므로, 묻지 못한 것과 데이터가 없는 것을 같은 값으로 두지 않는다.
            logger.error({ err, stockCode, minuteInterval }, '[KISCandleService] 분봉 캔들 조회 실패');
            throw err;
        }
    }

    /**
     * 1분봉 데이터를 N분봉으로 리샘플링
     * 오름차순 정렬된 OHLCV 배열 기준
     */
    resampleMinuteCandles(candles: number[][], intervalMinutes: number): number[][] {
        return resampleCandles(candles, intervalMinutes);
    }

    /**
     * 해외주식 기간별 캔들 조회 (HHDFS76240000)
     *
     * KIS 한 번 호출당 100건 반환 — 더 필요하면 BYMD 를 이전 페이지 마지막 일자로 갱신해
     * 반복 호출. timeframe 은 1d/1w/1M 만 지원 (해외 분봉은 `kis.fetchOverseasMinuteOHLCV` 가 준다).
     * BYMD 는 미국 거래일이라 실행 환경의 시간대가 아니라 미국 동부 날짜로 적는다.
     * `since <= 시각 <= until` 인 봉을 ccxt 규칙대로 `limit` 개 돌려준다. `since` 가 있으면 그 시각에 닿을 때까지 넘긴다.
     */
    async fetchOverseasDailyOHLCV(
        ticker: string,
        market: OverseasMarket,
        timeframe: string,
        limit: number,
        since?: number,
        until?: number,
    ): Promise<number[][]> {
        const gubn = OVERSEAS_GUBN_MAP[timeframe];
        const periodDays = OVERSEAS_PERIOD_DAYS[timeframe];
        if (!gubn || periodDays === undefined) {
            logger.warn({ ticker, market, timeframe },
                '[KISCandleService] 해외 timeframe 미지원 — 1d/1w/1M 만 사용 가능');
            return [];
        }

        try {
            const all: OhlcvRow[] = [];
            // `since` 가 있으면 첫 기준일을 `since` 에서 `limit` 개를 덮는 날로 당긴다.
            let end = until ?? this.exchange.milliseconds();
            if (since !== undefined) {
                end = Math.min(end, since + (limit * periodDays * DATE_MARGIN_FACTOR + OVERSEAS_END_PAD_DAYS) * DAY_MS);
            }
            let bymd = etYmd(end);
            let reachedSince = false;
            let exhausted = false;
            const PAGE_SIZE = 100;
            const MAX_PAGES = 10;

            for (let page = 0; page < MAX_PAGES && (since === undefined ? all.length < limit : !reachedSince); page++) {
                const response = await this.exchange.privateGetUapiOverseasPriceV1QuotationsDailyprice({
                    AUTH: '',
                    EXCD: market,
                    SYMB: ticker.toUpperCase(),
                    GUBN: gubn,
                    BYMD: bymd,
                    MODP: '1',
                    tr_id: CANDLE_TR_IDS.OVERSEAS_DAILY_CHART,
                });
                const data = response.output2 as KisOverseasDailyCandle[] | undefined;

                if (!Array.isArray(data) || data.length === 0) {
                    exhausted = true;
                    break;
                }

                for (const c of data) {
                    if (!c.xymd) continue;
                    const dateStr = c.xymd;
                    const ts = new Date(
                        `${dateStr.slice(0, 4)}-${dateStr.slice(4, 6)}-${dateStr.slice(6, 8)}T00:00:00Z`,
                    ).getTime();
                    if (since !== undefined && ts <= since) reachedSince = true;
                    all.push([
                        ts,
                        Number(c.open),
                        Number(c.high),
                        Number(c.low),
                        Number(c.clos),
                        Number(c.tvol),
                    ]);
                }

                const lastXymd = data[data.length - 1]?.xymd;
                if (data.length < PAGE_SIZE || !lastXymd) {
                    exhausted = true;
                    break;
                }
                // 다음 페이지: 마지막 일자보다 하루 전을 BYMD 로
                const lastDate = new Date(
                    `${lastXymd.slice(0, 4)}-${lastXymd.slice(4, 6)}-${lastXymd.slice(6, 8)}T00:00:00Z`,
                );
                lastDate.setUTCDate(lastDate.getUTCDate() - 1);
                bymd = this.formatUtcDate(lastDate);
            }

            if (since !== undefined && !reachedSince && !exhausted) {
                logger.warn({ ticker, market, timeframe, since, pages: MAX_PAGES },
                    '[KISCandleService] 페이지 상한에 걸려 since 까지 거슬러 가지 못했다');
            }
            const sorted = all.sort((a, b) => a[0] - b[0]);
            logger.info({ ticker, market, timeframe, count: sorted.length },
                '[KISCandleService] fetchOverseasDailyOHLCV 완료');
            return sliceCandleWindow(sorted, since, until, limit);
        } catch (err) {
            logger.error({ err, ticker, market, timeframe },
                '[KISCandleService] 해외 기간별 캔들 조회 실패');
            throw err;
        }
    }

    /** 날짜를 UTC 달력 기준 YYYYMMDD 로 변환 */
    private formatUtcDate(date: Date): string {
        const y = date.getUTCFullYear();
        const m = String(date.getUTCMonth() + 1).padStart(2, '0');
        const d = String(date.getUTCDate()).padStart(2, '0');
        return `${y}${m}${d}`;
    }
}

/** @deprecated `KisCandleService` 를 쓴다. 다음 판에서 지운다. */
export const KISCandleService = KisCandleService;
/** @deprecated `KisCandleService` 를 쓴다. 다음 판에서 지운다. */
export type KISCandleService = KisCandleService;

File: python/kr_broker/kis_candle_service.py
# 이 파일은 scripts/gen-python-sync.mjs 가 python/kr_broker/async_support/kis_candle_service.py 에서 만든다. 직접 고치지 않는다.
"""한국투자증권이 직접 주는 봉(OHLCV). TypeScript 판 `ts/src/kis/kis-candle-service.ts` 와 같다.

`kis.fetch_ohlcv` 는 이력이 긴 야후를 먼저 쓴다. 이 모듈은 KIS 원본 봉이 필요한 경로(깊은 이력 채우기, 당일 분봉, 야후가 빈 해외 일봉)를
`kis` 인스턴스의 암묵 API 로 조회한다. 인스턴스의 `candles()` 가 이 서비스를 돌려준다.

국내 날짜 인자(`YYYYMMDD`)는 실행 환경의 시간대가 아니라 한국 날짜로 적는다.
해외 기간별 시세의 기준일(`BYMD`)은 미국 거래일이라 미국 동부 날짜로 적는다.
"""

import datetime
import logging
import math
from typing import Any, Callable, Dict, List, Optional

from kr_broker.base import functions as fn
from kr_broker.kis_candle_pagination import merge_candles, plan_windows, slice_candle_window, to_kis_date
from kr_broker.kis_candle_resample import resample_candles
from kr_broker.us_market_hours import et_ymd

logger = logging.getLogger('kr_broker')

CANDLE_TR_IDS = {
    # 국내 기간별 시세(일·주·월·년)
    'DAILY_CHART': 'FHKST03010100',
    # 국내 당일 분봉
    'MINUTE_CHART': 'FHKST03010200',
    # 해외 기간별 시세(일·주·월)
    'OVERSEAS_DAILY_CHART': 'HHDFS76240000',
}
# 해외 일·주·월 구분(GUBN).
OVERSEAS_GUBN_MAP = {'1d': '0', '1w': '1', '1W': '1', '1M': '2'}
# 분봉 API 한 번의 최대 행 수.
MINUTE_PAGE_SIZE = 30
# 기간 코드별 일수 승수와 주말·휴장일 여유.
PERIOD_DAY_MULTIPLIER = {'W': 7, 'M': 30}
DATE_MARGIN_FACTOR = 1.5
# 해외 일·주·월봉 한 개가 차지하는 달력일 수와, `since` 에서 `limit` 개를 덮는 기준일을 어림할 때 더하는 달력일(연휴를 덮는다).
OVERSEAS_PERIOD_DAYS = {'1d': 1, '1w': 7, '1W': 7, '1M': 31}
OVERSEAS_END_PAD_DAYS = 7
# 해외 기간별 시세 한 번의 행 수와 페이지 상한.
OVERSEAS_PAGE_SIZE = 100
OVERSEAS_MAX_PAGES = 10
DAY_MS = 24 * 60 * 60 * 1000


def _format_utc_date(ms: float) -> str:
    """UTC 밀리초의 UTC 달력 날짜 `YYYYMMDD`. 시각이 아니면 JavaScript 처럼 `NaNNaNNaN` 이다."""
    if not isinstance(ms, (int, float)) or not math.isfinite(ms):
        return 'NaNNaNNaN'
    d = datetime.datetime(1970, 1, 1) + datetime.timedelta(milliseconds=ms)
    return f'{d.year}{d.month:02d}{d.day:02d}'


def _number(row: Dict[str, Any], key: str) -> float:
    """JavaScript 의 `Number(row.key)`. 키가 없으면 NaN 이다."""
    return fn.js_number(row[key]) if key in row else math.nan


def _parse_ms(text: str) -> float:
    """JavaScript 의 `new Date(text).getTime()`. 못 읽으면 NaN 이다."""
    parsed = fn.js_date_parse_iso(text)
    return math.nan if parsed is None else parsed


def _output2(response: Any) -> Any:
    return response.get('output2') if isinstance(response, dict) else None


class KISCandleService:
    """KIS 원본 봉 조회. `exchange` 는 자격증명을 채운 `kis` 인스턴스다."""

    def __init__(self, exchange: Any) -> None:
        self.exchange = exchange

    def fetch_daily_ohlcv(self, stock_code: str, period_code: str, limit: int) -> List[List[float]]:
        """국내 기간별(일·주·월) 봉. 오늘(한국 날짜)로 끝나는 기간을 계산해 부르므로 한 번에 100행쯤까지다."""
        now = self.exchange.milliseconds()
        multiplier = PERIOD_DAY_MULTIPLIER.get(period_code, 1)
        start_date = to_kis_date(now - limit * multiplier * DATE_MARGIN_FACTOR * DAY_MS)
        rows = self.fetch_daily_ohlcv_range(stock_code, period_code, start_date, to_kis_date(now))
        return rows[-limit:]

    def fetch_daily_ohlcv_paged(self, stock_code: str, period_code: str, needed_candles: float, now_ms: Optional[int] = None,
                                on_page: Optional[Callable[[int], None]] = None) -> List[List[float]]:
        """창을 과거로 옮겨 가며 여러 번 불러 깊은 이력을 모은다. 빈 창이 나오면(상장 전) 멈춘다. `needed_candles` 가 0 이하면 부르지 않는다."""
        windows = plan_windows(needed_candles, self.exchange.milliseconds() if now_ms is None else now_ms)
        if len(windows) == 0:
            return []
        pages: List[List[List[float]]] = []
        for i, window in enumerate(windows):
            rows = self.fetch_daily_ohlcv_range(stock_code, period_code, window['start'], window['end'])
            if len(rows) == 0:
                logger.debug('[KISCandleService] 빈 창 — 페이지네이션 중단(상장 이전 추정) (stockCode=%s, page=%d)', stock_code, i + 1)
                break
            pages.append(rows)
            if on_page is not None:
                on_page(i)
        merged = merge_candles(pages)
        logger.info('[KISCandleService] 페이지네이션 완료 (stockCode=%s, periodCode=%s, pages=%d, candles=%d)',
                    stock_code, period_code, len(pages), len(merged))
        return merged

    def fetch_daily_ohlcv_range(self, stock_code: str, period_code: str, start_date: str, end_date: str) -> List[List[float]]:
        """기간(`YYYYMMDD`)을 정해 국내 일·주·월 봉을 받는다. 한 응답이 100행쯤이다. 실패하면 로그를 남기고 던진다."""
        try:
            response = self.exchange.private_get_uapi_domestic_stock_v1_quotations_inquire_daily_itemchartprice({
                'FID_COND_MRKT_DIV_CODE': 'J',
                'FID_INPUT_ISCD': stock_code,
                'FID_INPUT_DATE_1': start_date,
                'FID_INPUT_DATE_2': end_date,
                'FID_PERIOD_DIV_CODE': period_code,
                'FID_ORG_ADJ_PRC': '0',
                'tr_id': CANDLE_TR_IDS['DAILY_CHART'],
            })
            data = _output2(response)
            logger.info('[KISCandleService] fetchDailyOHLCV 응답 (stockCode=%s, periodCode=%s, %s~%s, length=%s)', stock_code, period_code,
                        start_date, end_date, len(data) if isinstance(data, list) else 'N/A')
            if not isinstance(data, list):
                return []
            candles = []
            for candle in data:
                date_str = candle['stck_bsop_date']
                timestamp = _parse_ms(f'{date_str[0:4]}-{date_str[4:6]}-{date_str[6:8]}T09:00:00+09:00')
                candles.append([timestamp, _number(candle, 'stck_oprc'), _number(candle, 'stck_hgpr'), _number(candle, 'stck_lwpr'),
                                _number(candle, 'stck_clpr'), _number(candle, 'acml_vol')])
            return sorted(candles, key=lambda c: c[0])
        except Exception:
            # 실패를 빈 목록으로 바꾸지 않는다(분봉과 같다). 빈 창은 페이지네이션이 "상장 이전"으로 읽는다.
            logger.error('[KISCandleService] 기간별 캔들 조회 실패 (stockCode=%s, periodCode=%s, %s~%s)', stock_code, period_code, start_date,
                         end_date, exc_info=True)
            raise

    def fetch_minute_ohlcv(self, stock_code: str, minute_interval: int, limit: int) -> List[List[float]]:
        """국내 당일 분봉. 한 번에 30건씩 이어서 `limit` 개까지 모은다. 실패는 빈 목록으로 바꾸지 않고 던진다.
        빈 목록은 휴장·거래정지로 읽히므로, 묻지 못한 것과 데이터가 없는 것을 같은 값으로 두지 않는다."""
        try:
            # 연속조회는 커서 시각의 봉을 다음 쪽에 다시 줄 수 있어, 시각마다 한 봉만 담는다.
            by_timestamp: Dict[float, List[float]] = {}
            cursor = ''
            while len(by_timestamp) < limit:
                response = self.exchange.private_get_uapi_domestic_stock_v1_quotations_inquire_time_itemchartprice({
                    'FID_COND_MRKT_DIV_CODE': 'J',
                    'FID_INPUT_ISCD': stock_code,
                    'FID_INPUT_HOUR_1': cursor,
                    'FID_PW_DATA_INCU_YN': 'N',
                    'FID_ETC_CLS_CODE': '',
                    'tr_id': CANDLE_TR_IDS['MINUTE_CHART'],
                })
                data = _output2(response)
                if not isinstance(data, list) or len(data) == 0:
                    break
                for item in data:
                    date_str = item.get('stck_bsop_date')
                    time_str = item.get('stck_cntg_hour')
                    if not date_str or not time_str:
                        continue
                    timestamp = _parse_ms(f'{date_str[0:4]}-{date_str[4:6]}-{date_str[6:8]}'
                                          f'T{time_str[0:2]}:{time_str[2:4]}:{time_str[4:6]}+09:00')
                    by_timestamp[timestamp] = [timestamp, _number(item, 'stck_oprc'), _number(item, 'stck_hgpr'),
                                               _number(item, 'stck_lwpr'), _number(item, 'stck_prpr'), _number(item, 'cntg_vol')]
                if len(data) < MINUTE_PAGE_SIZE:
                    break
                last_time = data[-1].get('stck_cntg_hour')
                if not last_time or last_time == cursor:
                    break
                cursor = last_time
            logger.info('[KISCandleService] fetchMinuteOHLCV 완료 (stockCode=%s, minuteInterval=%s, count=%d)', stock_code, minute_interval,
                        len(by_timestamp))
            ordered = sorted(by_timestamp.values(), key=lambda c: c[0])
            if minute_interval <= 1:
                return ordered[-limit:]
            return self.resample_minute_candles(ordered, minute_interval)[-limit:]
        except Exception:
            logger.error('[KISCandleService] 분봉 캔들 조회 실패 (stockCode=%s, minuteInterval=%s)', stock_code, minute_interval, exc_info=True)
            raise

    def resample_minute_candles(self, candles: List[List[float]], interval_minutes: int) -> List[List[float]]:
        """1분봉을 N분봉으로 합친다. 입력은 시각 오름차순이다."""
        return resample_candles(candles, interval_minutes)

    def fetch_overseas_daily_ohlcv(self, ticker: str, market: str, timeframe: str, limit: int, since: Optional[int] = None,
                                   until: Optional[int] = None) -> List[List[float]]:
        """해외 기간별(일·주·월) 봉(`HHDFS76240000`). 한 번에 100건이라, 더 필요하면 기준일(`BYMD`)을 앞 페이지 마지막 날의 하루 전으로
        옮겨 다시 부른다. `1d`·`1w`·`1M` 만 받고 다른 봉은 빈 목록이다(KIS 해외 분봉은 이 경로에 없다). 실패하면 로그를 남기고 던진다.
        `BYMD` 는 미국 거래일이라 실행 환경의 시간대가 아니라 미국 동부 날짜로 적는다.
        `since <= 시각 <= until` 인 봉을 ccxt 규칙대로 `limit` 개 돌려준다. `since` 가 있으면 그 시각에 닿을 때까지 넘긴다."""
        gubn = OVERSEAS_GUBN_MAP.get(timeframe)
        if not gubn:
            logger.warning('[KISCandleService] 해외 timeframe 미지원 — 1d/1w/1M 만 사용 가능 (ticker=%s, market=%s, timeframe=%s)',
                           ticker, market, timeframe)
            return []
        try:
            collected: List[List[float]] = []
            # `since` 가 있으면 첫 기준일을 `since` 에서 `limit` 개를 덮는 날로 당긴다.
            end = self.exchange.milliseconds() if until is None else until
            if since is not None:
                end = min(end, since + (limit * OVERSEAS_PERIOD_DAYS[timeframe] * DATE_MARGIN_FACTOR + OVERSEAS_END_PAD_DAYS) * DAY_MS)
            bymd = et_ymd(end)
            reached_since = False
            exhausted = False
            page = 0
            while page < OVERSEAS_MAX_PAGES and (len(collected) < limit if since is None else not reached_since):
                response = self.exchange.private_get_uapi_overseas_price_v1_quotations_dailyprice({
                    'AUTH': '',
                    'EXCD': market,
                    'SYMB': ticker.upper(),
                    'GUBN': gubn,
                    'BYMD': bymd,
                    'MODP': '1',
                    'tr_id': CANDLE_TR_IDS['OVERSEAS_DAILY_CHART'],
                })
                data = _output2(response)
                if not isinstance(data, list) or len(data) == 0:
                    exhausted = True
                    break
                for c in data:
                    date_str = c.get('xymd')
                    if not date_str:
                        continue
                    ts = _parse_ms(f'{date_str[0:4]}-{date_str[4:6]}-{date_str[6:8]}T00:00:00Z')
                    if since is not None and ts <= since:
                        reached_since = True
                    collected.append([ts, _number(c, 'open'), _number(c, 'high'), _number(c, 'low'), _number(c, 'clos'), _number(c, 'tvol')])
                last_xymd = data[-1].get('xymd')
                if len(data) < OVERSEAS_PAGE_SIZE or not last_xymd:
                    exhausted = True
                    break
                # 다음 페이지는 마지막 날의 하루 전을 달력 날짜 그대로(UTC) 적어 부른다.
                last_date = _parse_ms(f'{last_xymd[0:4]}-{last_xymd[4:6]}-{last_xymd[6:8]}T00:00:00Z')
                bymd = _format_utc_date(last_date - DAY_MS)
                page += 1
            if since is not None and not reached_since and not exhausted:
                logger.warning('[KISCandleService] 페이지 상한에 걸려 since 까지 거슬러 가지 못했다 '
                               '(ticker=%s, market=%s, timeframe=%s, since=%s, pages=%d)', ticker, market, timeframe, since,
                               OVERSEAS_MAX_PAGES)
            ordered = sorted(collected, key=lambda c: c[0])
            logger.info('[KISCandleService] fetchOverseasDailyOHLCV 완료 (ticker=%s, market=%s, timeframe=%s, count=%d)', ticker, market,
                        timeframe, len(ordered))
            return slice_candle_window(ordered, since, until, limit)
        except Exception:
            logger.error('[KISCandleService] 해외 기간별 캔들 조회 실패 (ticker=%s, market=%s, timeframe=%s)', ticker, market, timeframe,
                         exc_info=True)
            raise
